### PR TITLE
Storm-697: Support for Emitting Kafka Message Offset and Partition

### PR DESCRIPTION
**Pull request by mjtieman**
It would be nice expose the offset and partition of messages consumed from Kafka to the Scheme generating the Tuples. This is useful for auditing/replaying data from arbitrary points on a Kafka topic, saving the partition and offset of each message of a discrete stream instead of persisting the entire message.

* Added new sheme to that accepts a Partition and the message offset in its deserialization method.
* Defined an overload of `KafkaUtils.generateTuples` to accept a `Partition` and offset in addition to the message byte[].
* Added a flag in `KafkaConfig` to indicate if the metadata (partition and offset) should be available during tuple generation.
* Wrote a simple String implementation of the new scheme, `StringMessageAndMetadataScheme`, following the same pattern as `StringKeyValueScheme`.
* Unit tests.

### CODE REVIEW

**Comment by msaunier-poctu** [comment]
:+1:

TridentKafkaEmitter should be updated to support emitting offset in Trident :


```diff
diff --git a/external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaEmitter.java b/external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaEmitter.java
index 94bf134..dc9bb6d 100644
--- a/external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaEmitter.java
+++ b/external/storm-kafka/src/jvm/storm/kafka/trident/TridentKafkaEmitter.java
@@ -113,7 +113,7 @@ public class TridentKafkaEmitter {
         ByteBufferMessageSet msgs = fetchMessages(consumer, partition, offset);
         long endoffset = offset;
         for (MessageAndOffset msg : msgs) {
-            emit(collector, msg.message());
+            emit(collector, msg.message(),partition,msg.offset());
             endoffset = msg.nextOffset();
         }
         Map newMeta = new HashMap();
@@ -160,14 +160,19 @@ public class TridentKafkaEmitter {
                 if (offset > nextOffset) {
                     throw new RuntimeException("Error when re-emitting batch. overshot the end offset");
                 }
-                emit(collector, msg.message());
+                emit(collector, msg.message(),partition,msg.offset());
                 offset = msg.nextOffset();
             }
         }
     }
 
-    private void emit(TridentCollector collector, Message msg) {
-        Iterable<List<Object>> values = KafkaUtils.generateTuples(_config, msg);
+    private void emit(TridentCollector collector, Message msg, Partition partition, long offset) {
+        Iterable<List<Object>> values;
+        if(_config.tupleMetaData) {
+            values = KafkaUtils.generateTuples(_config, msg, partition, offset);
+        }else{
+            values = KafkaUtils.generateTuples(_config, msg);
+        }
         if (values != null) {
             for (List<Object> value : values) {
                 collector.emit(value);
```

**Comment by mjtieman** [comment]
Good call, I will update TridentKafkaEmitter as well.

**Comment by mjtieman** [comment]
Done.

**Comment by lbruand-poctu** [comment]
:+1:

**Comment by mjtieman** [comment]
If there are no other comments on these changes let's get them merged :smile: .

**Comment by lbruand-poctu** [comment]
Hi!

What is in the way of the merging of these changes ?
When can we expect them to be merged ?

Cheers,
Lucas

**Comment by mjtieman** [comment]
I have never merged into storm before, reading on how to perform the merge.

**Comment by mjtieman** [comment]
Apologies, I have to register as an Apache committer before I can merge these changes.

**Comment by mjtieman** [comment]
Again apologies, I have never contributed code to an open source project before. Clearly I did not understand the distinction between a contributor (me) and a committer. Is there a formal way to request a pull request be merged by a committer?

**Comment by HeartSaVioR** [comment]
@mjtieman 
No, you're going the right way.
You can refer http://storm.apache.org/documentation/BYLAWS.html to how project works.

**Comment by maverick2202** [comment]
Can we merge this code into 0.10 release ?

**Comment by mjtieman** [comment]
This branch has fallen a little behind since the changes have been made. I will resolve the conflicts this evening.

**Comment by maverick2202** [comment]
thanks.

**Comment by mjtieman** [comment]
Conflicts resolved.

**Comment by mjtieman** [comment]
Resolved conflicts in PartitionManager. Let me know if there is anything I can fix that is blocking these changes from being mergd.

**Comment by mjtieman** [comment]
@HeartSaVioR Mind taking a look at these changes to determine if they should be merged?

**Comment by HeartSaVioR** [comment]
@mjtieman 
Sorry for the long delay I didn't catch this before.

One minor thing, seems like we don't need tupleMetaData flag.
We can know what information such user want to take, by only checking scheme type, and this PR already does it.
When it doesn't use StringMessageAndMetadataScheme, tupleMetaData flag is no effect. I think it shows flag is not needed.

What do you think?

**Comment by HeartSaVioR** [comment]
Otherwise looks good overall.

**Comment by mjtieman** [comment]
@HeartSaVioR No worries. I agree that the tupleMetaData flag is superfluous. The instanceof check that is being done to guarantee type safety can be used to determine which path to take when generating the tuples.

This will require a little bit of tweaking since we will be moving the scheme instanceof check and cast out of KafkaUtils and into PartitionManager. Updating now.

**Comment by mjtieman** [comment]
Done.

**Comment by HeartSaVioR** [comment]
+1

**Comment by HeartSaVioR** [comment]
@mjtieman 
Could you add this feature to external/storm-kafka/README.md so that users can see it? Thanks!

**Comment by mjtieman** [comment]
Will do. I will also update the CHANGELOG and contributors section of the README.

**Comment by HeartSaVioR** [comment]
You don't need to update CHANGELOG / README since it could increase probability of confliction.

**Comment by mjtieman** [comment]
Updated the storm-kafka to list MessageMetadataSchemeAsMultiScheme as a MultiScheme implementation. Let me know if there is anything else I should add.

File: external/storm-kafka/src/jvm/storm/kafka/MessageMetadataScheme.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package storm.kafka;
+
+import java.util.List;
+import backtype.storm.spout.Scheme;
+
+/**
```
Comment (HeartSaVioR) [inline external/storm-kafka/src/jvm/storm/kafka/MessageMetadataScheme.java]: License header should be put to the first line of the file.

File: external/storm-kafka/src/jvm/storm/kafka/MessageMetadataSchemeAsMultiScheme.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package storm.kafka;
```
Comment (HeartSaVioR) [inline external/storm-kafka/src/jvm/storm/kafka/MessageMetadataSchemeAsMultiScheme.java]: License header should be put to the first line of the file.

File: external/storm-kafka/src/jvm/storm/kafka/StringMessageAndMetadataScheme.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package storm.kafka;
```
Comment (HeartSaVioR) [inline external/storm-kafka/src/jvm/storm/kafka/StringMessageAndMetadataScheme.java]: License header should be put to the first line of the file.

**Comment by HeartSaVioR** [comment]
I'd like to see explanation of MessageMetadataSchemeAsMultiScheme to README.md, since end users would like to see only README.md and play with it.
And I've found that some files' license header are missing or wrongly placed.

Could you address it? Thanks!

**Comment by mjtieman** [comment]
Will do.

**Comment by mjtieman** [comment]
Done.

**Comment by HeartSaVioR** [comment]
+1

**Pull request merged**